### PR TITLE
Add support for moving `/push_rules` off of main process

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

**Review by sandhose — CHANGES_REQUESTED** [review]
I think I learned how replication usually works reading this 🙇 

If my understanding is right, this is to allow routing `/push_rules` to a separate worker.
Problem is, one other case where push rules are being written outside of the `/push_rules`, which is when we copy push rules when we upgrade rooms.
So this adds a replication handler for that particular operation, which is called if the room upgrade doesn't happen on the worker which is supposed to handle push rules

At least, this is what I understood reading the patch, so if that's completely not the case, either the patch is wrong or I can't read code anymore

File: synapse/config/workers.py [hunk 1]
```diff
@@ -182,6 +184,10 @@ class WriterLocations:
         default=["master"],
         converter=_instance_to_list_converter,
     )
+    push: List[str] = attr.ib(
+        default=["master"],
+        converter=_instance_to_list_converter,
+    )
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/config/workers.py:190]: Good point! Done.

**Comment by erikjohnston** [comment]
> I think I learned how replication usually works reading this 🙇
> 
> If my understanding is right, this is to allow routing `/push_rules` to a separate worker. Problem is, one other case where push rules are being written outside of the `/push_rules`, which is when we copy push rules when we upgrade rooms. So this adds a replication handler for that particular operation, which is called if the room upgrade doesn't happen on the worker which is supposed to handle push rules
> 
> At least, this is what I understood reading the patch, so if that's completely not the case, either the patch is wrong or I can't read code anymore

Yes, exactly

**Review by sandhose — APPROVED** [review]

File: docs/workers.md [hunk 2]
```diff
@@ -532,6 +532,13 @@ the stream writer for the `presence` stream:
 
     ^/_matrix/client/(api/v1|r0|v3|unstable)/presence/
 
+##### The `push` stream
+
+The following endpoints should be routed directly to the worker configured as
+the stream writer for the `push` stream:
+
+    ^/_matrix/client/(api/v1|r0|v3|unstable)/push_rules/
```
Comment (sandhose) [inline docs/workers.md:540]: WAIT it's actually `/pushrules/`
```suggestion
    ^/_matrix/client/(api/v1|r0|v3|unstable)/pushrules/
```

Comment (erikjohnston) [reply docs/workers.md:540]: Oh good spot

**Pull request merged**